### ISSUE TITLE
[DOC] Install information for Jetson Nano

### ISSUE THREAD

**Issue by vielmetti** [rapidsai/cusignal#8]
I think it might be possible to get this code running on the Jetson Nano, but I'm not able to at the moment.

The `conda` version for the Nano that I'm trying is Archiconda, https://github.com/Archiconda .

The Nano has Maxwell cores, which should be enough for this. But I don't think all the necessary dependencies have been put together in Archiconda, and I'm not exactly sure where to start. I get a series of `PackagesNotFoundError` messages saying that I can't install from current channels.

Before I go too much further, happy to provide more details as requested, or to understand whether this is intended to work or known not to work for some more fundamental reason!

**Comment by awthomp** [comment]
Welcome, @vielmetti and thanks for submitting an issue! cuSignal has been tested on a Jetson Nano and does indeed work.

There are a few core dependencies for cuSignal. 
* Numba
* NumPy
* SciPy
* cudatoolkit
* CuPy

If you look at the [cusignal_conda_env.yml](https://github.com/rapidsai/cusignal/blob/master/cusignal_conda_env.yml) file, there are others listed. I'm going to mark this issue as "good first issue" because we can do some dependency cleanup. cuDF, cuML, cuGraph, and PyTorch, among others, aren't "core dependencies" but are included for ecosystem compatibility. We are also working on native conda packaging that should simplify all of this!

On to getting cuSignal to work on a Nano:

I haven't tested Archiconda, but I was able to get the above dependencies installed with conda4aarch64. You can follow the install guide listed [here](https://numba.pydata.org/numba-doc/dev/user/installing.html#installing-on-linux-armv8-aarch64-platforms). 

Once conda4aarch64 (and Numba) are installed by following that link, you can proceed to follow the following instructions.

```
conda install numpy scipy cudatoolkit pip
pip install cupy-cudaXXX
python setup.py install
```

Note: the XXX above will be the version of the cudatoolkit you install. You can always specify this in in the conda install line, e.g. `conda install cudatoolkit=10.1`. Here, your cupy install command would be: `pip install cupy-cuda101`. CUDA 10.0 would yield `pip install cupy-cuda100`.

Let me know how this works, and if you run into any issues!

For contributions:
1. We should remove pytorch, cudf, cuml, cugraph, pytorch, and pandas from the core dependencies and document that they're optional for ecosystem completeness. This should also allow the conda env install to work on both Windows and Linux.
2. Document specific instructions for the Jetson embedded devices.

**Comment by vielmetti** [comment]
Thanks for the update! Here's how far I got.

1. conda4aarch64 install OK.
2. numba install OK.
3. numpy, scipy, pip install OK.

However, when I get to installing `cudatoolkit` via conda, I hit a snag: none of the available channels have a `cudatoolkit` for aarch64. 

https://anaconda.org/search?q=cudatoolkit

**Comment by awthomp** [comment]
@vielmetti Since you're running JetPack on the Nano, CUDA is already installed. You can find the version via:

`nvcc --version` or `cat /usr/local/cuda/version.txt`

I'd proceed like you installed the CUDA toolkit (the Conda version, as far as I know, just makes the libs local to your conda environment). You should be good to `pip install cupy-cudaXXX`, based on the version of CUDA you have.

More info on the CuPy install can be found [here](https://docs-cupy.chainer.org/en/stable/install.html).

**Comment by vielmetti** [comment]
Terrific! More progress.

I ended up installing `CuPy` from source, based on the instructions on the install you provided; `pip` didn't find a version, but the compile was fast enough.

I also installed `pytest` with `conda install pytest`, because otherwise you get the system version.

The whole `pytest` runs on this system:

123 passed, 57 warnings in 467.25s (0:07:47).  However it's not a completely quiet system (there's another job chewing up a CPU) so I will quiet it down to study it later.

Hopeful that this helps the next person along - thanks for all your assistance.

**Comment by vielmetti** [comment]
in this

https://github.com/rapidsai/cusignal/blob/master/notebooks/sdr_wfm_demod.ipynb

there's also a `SoapySDR` module -- perhaps @luigifreitas has some recommendations for installing that module inside conda, (cross reference to https://github.com/pothosware/SoapySDR/issues/251 )

**Comment by vielmetti** [comment]
and while I'm documenting things and they are all here in the same place --

At least a few of the examples look like they will have to be rewritten to run correctly on the Jetson Nano, not because the compute elements are wrong, but  because they run out of memory or perhaps some other resource on the (relatively memory constrained) Jetson Nano.

One error e.g. is `MemoryError: std::bad_alloc` when I run "Resample" from signaltools_examples. By tweaking the examples to use an order or two of magnitude less samples (1e6 instead of 1e8) they seem to run fine.

**Comment by awthomp** [comment]
I appreciate you working through getting cuSignal up and running on the Nano, @vielmetti. These are all great suggestions for documentation improvement in the future! Happy New Year!

**Comment by taureandyernv** [comment]
@awthomp and @vielmetti this would be a really cool "how-to" guide and blog!

**Comment by luigifreitas** [comment]
@vielmetti Sure! I'll look into a Conda based SoapySDR as soon as I get my Jetson Nano.

**Comment by vielmetti** [comment]
@taureandyernv agreed this would be a good writeup! I'd like to push through all the way to an application note for software-defined radio using cuSignal on the Nano, since I'm super curious about just what kind of results are possible with @luigifreitas code.

**Comment by evanmayer** [comment]
To add another data point here, I also ran into memory issues running the examples with the default values on a GTX 960. They went away with reduced array sizes, but I wasn't able to root cause exactly which size was the limit and why.

**Comment by awthomp** [comment]
Fixed in https://github.com/rapidsai/cusignal/pull/9

**Comment by vielmetti** [comment]
Thanks for the docs update. 

I did a brief writeup here

https://vielmetti.typepad.com/w8emv/2020/01/work-in-progress-cusignal-for-accelerated-signals-processing.html

trying to pack all of the relevant pieces why someone might be interested into a few words, and referring back to the docs and release here, the `conda4aarch64` install instructions, and @luigifreitas demos to date and future plans. (Happy to edit in any additions or corrections.)

**Comment by G4GUO** [comment]
I know this issue has been closed bu I have been trying to install it on a Xavier. I get as far as running pytest and it takes 285.26 secs to run.

I am now trying to run the E2E_Example. I had to downgrade numba==0.48.0 to get the first part to work. Now I have hit a problem with importing torch. This is due to there being no cudatoolkit package. I tried following the NVIDIA instructions
https://forums.developer.nvidia.com/t/pytorch-for-jetson-nano-version-1-4-0-now-available/72048
and try as I can I get unsupported platform messages. I am not really sure where to go from here.

- Charles

**Comment by awthomp** [comment]
Hey @G4GUO -- Have you followed the manual build instructions on the link you sent? The default conda Python environment is 3.7, so maybe specifying Python 3.6 will fix the issue?

Also see: https://elinux.org/Jetson_Zoo#PyTorch_.28Caffe2.29

**Comment by G4GUO** [comment]
Creating a new environment with python 3.6 fixed it.
I still had to downgrade numba also I get a warning about CuPY
"CuPy backend is only implemented for ndim == 1"
there are  a few other deprecated interface warnings
"Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument"
but it all runs now.
Even Xavier's fan comes on during training.

Next is to add the Soapy stuff (build Limesuite and copy across the required .py and .so files) and try that.

I think I am also going to need to write some cuda code later to generate the waterfall display but I 
will probably talk about that in a different thread.

- Charles

**Comment by awthomp** [comment]
Yay! Happy it worked for you, @G4GUO. I'll open an issue to get improved install documentation on the Jetson Devices. Ideally, we can get a conda package included so you don't have to go from source. I also need to build out the SoapySDR documentation. I was originally looking at getting that package into a conda channel too, but that's very preliminary work.

As a heads up, check out cusignal's spectrogram for your waterfall plot: https://docs.rapids.ai/api/cusignal/stable/api.html#cusignal.spectral.spectrogram

To display, you'll have to move the GPU array back to CPU. This can be done with `cupy.asnumpy()` https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.asnumpy.html

**Comment by G4GUO** [comment]
My plan is to do windowing / FFT / log mag / reduction to 8 bits / write a sliding buffer on the GPU which can then be displayed or chopped up into say 255x255 images for training. I did a GPU based LDPC decoder a while ago so I know how critical CUDA memory accesses are. I am a total novice at Python programming though and AI/ML. I have read a book about ML, watched a few TF videos and played with NVIDIA Digits and that is it. :)

- Charles

**Comment by awthomp** [comment]
@G4GUO -- Keep us updated on your work; we're happy to both help and expand the functionality of cuSignal. In the next few months, I hope to put up more examples for SDR applications and also the bridge to AI/ML techniques. The E2E notebook you mentioned needs a bit of updating but is hopefully a good place to start!

As for programming, we made the design decision in cusignal to avoid having a C++ layer and instead get GPU speedups on non-supported kernels with Numba and Raw CuPy CUDA kernels. A good example of how to do this is in `_upfirdn.py` located here: https://github.com/rapidsai/cusignal/blob/branch-0.14/python/cusignal/_upfirdn.py

Numba 1D implementation: https://github.com/rapidsai/cusignal/blob/f5db3abebd33693ed0aa92a625ad8b288893284f/python/cusignal/_upfirdn.py#L131

Raw CuPy CUDA Kernel, 1D: https://github.com/rapidsai/cusignal/blob/f5db3abebd33693ed0aa92a625ad8b288893284f/python/cusignal/_upfirdn.py#L173

When implementing something from scratch, I typically do: Python -> Numba -> Raw CuPy Kernel.

**Comment by luigifreitas** [comment]
@awthomp I would be happy to add an example using PyTorch with cuSignal! I'm using something similar on CyberRadio with my neural FM [demodulator](https://github.com/luigifreitas/CyberRadio/blob/CursedNet/src/main/python/cursednet.py) at this moment. I can port it to a Jupyter Notebook.

**Comment by awthomp** [comment]
I'd love that, @luigifreitas !

**Comment by QSatDT** [comment]
@luigifreitas  and @vielmetti I am posting here in case you have any updates or recommendations from personal experience on trying to use cusignal and soapysdr on Nvidia Nano, with conda4aarch64. I am trying my best to write my own first example of an RF receiver on an Nvidia Nano, but I don't seem to manage to install correctly soapysdr from source inside the conda environment. Any insight would be helpful, as I am very new to the SDR world. Thanks!

**Comment by G4GUO** [comment]
What I did was to compile and install Limesuite from source then Soapy (you have to do it in that order otherwise Soapy doesn't find the Lime system. Then I copied  SoapySDR.py and  _SoapySDR.so somewhere they could be found (in my case in the notebooks/sdr folder (yes I know bad practice)). I also copied  SoapySDRUtil over so I could verify all the parts worked. I am sure someone else can give you better advice.

- Charles

**Comment by awthomp** [comment]
Hey @QSatDT -- The Deepwave Digital folks have a good walkthrough on their website [here](http://docs.deepwavedigital.com/Tutorials/6_conda.html#install-airstack-drivers-in-conda-environment). They use their own Soapy driver, though -- so you'll need to keep this in mind if you don't have one of their AIR-T devices.

That said, the magic of what @G4GUO explains is in this code snippet. This is from Deepwave's Air-Stack drivers + Conda instructions located [here](http://docs.deepwavedigital.com/Tutorials/6_conda.html#install-airstack-drivers-in-conda-environment):

```
#!/bin/bash
CONDA_PACKAGE_PATH=$(python -c 'import site; print(site.getsitepackages()[0])')
# Create symbolic links for SoapySDR to the Conda environment and see it
SOAPY_INSTALL_DIR="/usr/lib/python3/dist-packages"
SOAPY_FILE_NAMES="_SoapySDR.cpython-36m-aarch64-linux-gnu.so SoapySDR.py"
for file_name in ${SOAPY_FILE_NAMES}; do
    src_file="${SOAPY_INSTALL_DIR}/${file_name}"
    if [[ -f ${src_file} ]]; then
        ln -s ${src_file} ${CONDA_PACKAGE_PATH}/${file_name}
        if [[ $? -eq 0 ]] ; then
            echo "Successfully created symbolic link:"
            echo "  ${src_file} -> "
            echo "  ${CONDA_PACKAGE_PATH}/${file_name}"
        fi
    else
        echo "Source file does not exist: ${src_file}"
        echo "Make sure SoapySDR drivers are properly installed."
    fi
done
export LD_LIBRARY_PATH=${SOAPY_INSTALL_DIR}
```

Long story short, you're doing exactly what Charles says. I personally installed SoapySDR on Ubuntu via apt-get and then copied the SoapySDR bindings from `/usr/lib/python3/site-packages` into my local conda environment. I also added `/usr/lib/python3/site-packages` to my `LD_LIBRARY_PATH`. There's some shell magic above, but this is the net-net.

**Comment by QSatDT** [comment]
@awthomp @G4GUO  Thank you very much for your kind replies! :) I will immediately proceed to test it out! Thank you also for the reference to DeepWave Digital, I find it very useful for what I am trying to learn.
